pallets/remarks/src/lib.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright 2023 Centrifuge Foundation (centrifuge.io).
+//
+// This file is part of the Centrifuge chain project.
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+mod weights;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::{dispatch::GetDispatchInfo, pallet_prelude::*, traits::IsSubType};
+	use frame_system::pallet_prelude::*;
+	use sp_runtime::traits::Dispatchable;
+	use sp_std::boxed::Box;
+
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The overarching call type.
+		type RuntimeCall: Parameter
+			+ Dispatchable<RuntimeOrigin = Self::RuntimeOrigin>
+			+ GetDispatchInfo
+			+ From<frame_system::Call<Self>>
+			+ IsSubType<Call<Self>>
+			+ IsType<<Self as frame_system::Config>::RuntimeCall>;
+
+		/// Weight information.
+		type WeightInfo: WeightInfo;
+
+		/// The type attached to the Remark Event.
+		type Remark: Parameter + Member;
+
+		/// Type that restrains the maximum remarks that can be attached to a
+		/// call.
+		type MaxRemarks: Get<u32>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A remark was made.
+		Remark {
+			remarks: BoundedVec<T::Remark, T::MaxRemarks>,
+			call: <T as Config>::RuntimeCall,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Add remarks to a call.
+		///
+		/// The weight calculation is similar to the one in the Proxy pallet.
+		#[pallet::call_index(0)]
+		#[pallet::weight({
+			let dispatch_info = call.get_dispatch_info();
+			(T::WeightInfo::remark()
+				.saturating_add(dispatch_info.weight),
+			dispatch_info.class)
+		})]
+		pub fn remark(
+			origin: OriginFor<T>,
+			remarks: BoundedVec<T::Remark, T::MaxRemarks>,
+			call: Box<<T as Config>::RuntimeCall>,
```

Comment:
Question:

How can we ensure that `call` is not a batch with other calls, i.e containing a remark call inside?

Reply:
Hmm, we could add some more trait constraints that would allow us to cast this into whatever calls that we want to check, but I'm not sure if we wanna do that. I guess this should be pretty simple and generic, and I doubt that the utility pallet is doing that for example.

Reply:
On second thought, not allowing a remark inside a remark makes sense, gonna look into adding that. Thank you!

Reply:
I added a filter, similar to the one in the utility pallet.

Reply:
Great! Thanks!!

I ask that because the reason to not have the utility pallet was precisely that you could add a remark in a batch, IIRC